weather-gov/weather.gov
# Initial i18n setup

## What does this PR do? 🛠️
This PR implements the recommendations in #1347 and is a follow-on to  the ADR in #1427
  
## What does the reviewer need to know? 🤔
  
### Basic settings ###
With the enabling of the Drupal core Language and Interface Translation modules (reflected in the exported config), we now have the capability to localize the site in multiple languages. Those languages need to be added in the "Languages" part of the administration settings. For now, we only have one language enabled (English).
  
(NOTE: We have enabled English-English translations for now, as a way of testing our system. However, if English is set as the default language we can also turn off the need to provide translations for it later. For now it is beneficial to keep on until we officially start working in a second language).
  
The name of the specific setting for translating static assets from templates or other config files is "Interface Translation," and the config updates in this PR should also account for the correct settings there. Note: we do _not_ want to update any individual phrase translations from within the CMS here, even though it offers that capability. We will strictly be using the `.po` files (see below) that are checked into this code repository.
  
### Translation files ###
Drupal uses the [gettext file format](https://www.gnu.org/software/gettext/manual/html_node/PO-Files.html) for specifying translations in each language. 
  
The convention is to name each file with the two-letter code for the given language, followed by the gettext file extension of `.po`. So for English, our translation file is named `en.po`.
  
Gettext uses the convention where the English phrase is itself the key to other translations. For example, a portion of a Spanish translation file might look something like:
```gettext
msgid "Hello"
msgstr "Hola"
```
  
Awkwardly, this means that our English translation file would have the same phrases as both the lookup and value. Fortunately, if we give a value of an empty string in these gettext files, the system will default to using the lookup key, which is already the English version of the phrase.
  
The initial translation files were generated first using the [`potx`](https://www.drupal.org/project/potx) tool, followed by removing some extraneous pairs by hand.
  
### New module `weather_i18n` ###
Following the example of the vote-gov team, our translation files as well as some i18n settings will be stored in a new custom module called `weather_i18n`. Translation files will be saved to `web/modules/weather_i18n/translations`. 
  
We are using a custom module in part to have a regular place to store the translation files, but also so that any edge cases we encounter down the road can be addressed by module hooks or similar (ths is what happened with vote-gov).
  
### Updating translations ###
If you edit a `.po` file, you will need to let Drupal know there are new translation updates, and also clear the cache. We have added a Makefile command `make ut` which will handle this for you.
  
## Future work ##
We need to address testing the consistency of these files in #1429 
  
## Try it ##
If you want to test this for yourself, try the following:
1. Pull down this branch, import the config, then clear cache or zap if needed
2. Open the english translation file and add a new English-to-English translation for "Day" (ie, edit the `msgstr` value where the `msgid` is "Day").
3. Run make ut
4. Load any forecast page, then look at the daily forecast cards. Instead of saying "Day" for the daytime segment, it should now have the translated string you edited.
